### PR TITLE
Clap v3 update: keygen

### PR DESCRIPTION
**Pull request by KirillLykov**
#### Problem

Update clap to version 3. This PR is part of the [effort](https://github.com/solana-labs/solana/issues/23557#issuecomment-1075335215) to migrate the whole repo to clap-v3

#### Summary of Changes

* update CLI to use clap v3
* switch from `clap-utils` to `clap-v3-utils`

### CODE REVIEW

File: keygen/src/keygen.rs [hunk 1]
```diff
@@ -350,101 +348,104 @@ fn main() -> Result<(), Box<dyn error::Error>> {
...
+                    Arg::new("starts_with")
                         .long("starts-with")
                         .value_name("PREFIX:COUNT")
                         .number_of_values(1)
                         .takes_value(true)
-                        .multiple(true)
-                        .validator(grind_validator_starts_with)
+                        .multiple_occurrences(true)
+                        .multiple_values(true)
+                        .validator(|s| grind_validator_starts_with(s.to_string()))
```
Comment (t-nelson) [inline keygen/src/keygen.rs:413]: doesn't look like `grind_validator_starts_with` strictly requires a `String`. I'd prefer to change its signature to take `&str`

File: keygen/src/keygen.rs [hunk 2]
```diff
@@ -350,101 +348,104 @@ fn main() -> Result<(), Box<dyn error::Error>> {
...
+                    Arg::new("ends_with")
                         .long("ends-with")
                         .value_name("SUFFIX:COUNT")
                         .number_of_values(1)
                         .takes_value(true)
-                        .multiple(true)
-                        .validator(grind_validator_ends_with)
+                        .multiple_occurrences(true)
+                        .multiple_values(true)
+                        .validator(|s| grind_validator_ends_with(s.to_string()))
```
Comment (t-nelson) [inline keygen/src/keygen.rs:424]: ditto

File: keygen/src/keygen.rs [hunk 3]
```diff
@@ -350,101 +348,104 @@ fn main() -> Result<(), Box<dyn error::Error>> {
...
                         .help("Saves specified number of keypairs whos public key starts and ends with the indicated perfix and suffix\nExample: --starts-and-ends-with sol:ana:4\nPREFIX and SUFFIX type is Base58\nCOUNT type is u64"),
                 )
                 .arg(
-                    Arg::with_name("num_threads")
+                    Arg::new("num_threads")
                         .long("num-threads")
                         .value_name("NUMBER")
                         .takes_value(true)
-                        .validator(is_parsable::<usize>)
+                        .validator(|s| is_parsable::<usize>(s.to_string()))
```
Comment (t-nelson) [inline keygen/src/keygen.rs:443]: what's constraining the v3 version of `is_parsable` to taking a `String`?

File: keygen/src/keygen.rs [hunk 4]
```diff
@@ -457,64 +458,64 @@ fn main() -> Result<(), Box<dyn error::Error>> {
...
-                .setting(AppSettings::DisableVersion)
+                .disable_version_flag(true)
                 .arg(
-                    Arg::with_name("prompt_signer")
+                    Arg::new("prompt_signer")
                         .index(1)
                         .value_name("KEYPAIR")
                         .takes_value(true)
-                        .validator(is_prompt_signer_source)
+                        .validator(|s| is_prompt_signer_source(s))
```
Comment (t-nelson) [inline keygen/src/keygen.rs:500]: superfluous closure?

File: keygen/src/keygen.rs [hunk 5]
```diff
@@ -535,7 +536,7 @@ fn do_main(matches: &ArgMatches<'_>) -> Result<(), Box<dyn error::Error>> {
     let mut wallet_manager = None;
 
     match matches.subcommand() {
```
Comment (t-nelson) [inline keygen/src/keygen.rs:538]: we can just `unwrap()`this, right?  clap will enforce the app definition and bail on unrecognized subcommands or flags before we get here.  probably cleaner than all of the `Some(...)`s

File: keygen/src/keygen.rs (hunk 4 above)
Comment (KirillLykov) [reply keygen/src/keygen.rs:500]: without this closure it doesn't compile with error:
```
error: implementation of `FnOnce` is not general enough
   --> keygen/src/keygen.rs:500:26
    |
500 |                         .validator(is_prompt_signer_source)
    |                          ^^^^^^^^^ implementation of `FnOnce` is not general enough
    |
    = note: `fn(&'2 str) -> Result<(), std::string::String> {is_prompt_signer_source::<&'2 str>}` must implement `FnOnce<(&'1 str,)>`, for any lifetime `'1`...
    = note: ...but it actually implements `FnOnce<(&'2 str,)>`, for some specific lifetime `'2`

error: could not compile `solana-keygen` due to previous error
```
I found explanation [here](https://users.rust-lang.org/t/implementation-of-fnonce-is-not-general-enough/68294/3).

File: keygen/src/keygen.rs (hunk 1 above)
Comment (KirillLykov) [reply keygen/src/keygen.rs]: thanks for pointing out, fixed all the occurrences except for one which is more elaborated

File: keygen/src/keygen.rs (hunk 4 above)
Comment (t-nelson) [reply keygen/src/keygen.rs:500]: even when adapting the signature of `is_prompt_signer_source` to explicitly take `&str`? I'd think it's just not able to resolve `AsRef<str>` to `&str` automatically :thinking:

Comment (KirillLykov) [reply keygen/src/keygen.rs]: If I use `&str` it works (updated). Looks like `validator` is always assuming `&str` so might be unnecessary to have generics there. My only concern is that all the functions in the `input_validators.rs` are using the same generic signature and may be there is a reason behind which is unknown to me. So I got rid of generic only in `is_prompt_signer_source`, others I left as they are to modify in next PRs by demand, if makes sense.

Comment (t-nelson) [reply keygen/src/keygen.rs]: cool! agreed updating all of the validators in a future PR is the right thing

**Review by t-nelson — APPROVED** [review]
lgtm. thanks!

**Pull request merged**